include/deal.II/fe/fe_coupling_values.h

```diff
@@ -0,0 +1,653 @@
+  RenumberedView<ViewType>::outer_to_inner_values(
+    std::vector<ValueType> &outer_values) const
+  {
+    AssertDimension(outer_values.size(), data.n_quadrature_points);
+    if (data.quadrature_renumbering.empty())
+      {
+        return outer_values;
+      }
+    else
+      {
+        const auto name = data.get_unique_container_name(
+          "RenumberedView::outer_to_inner_values",
+          data.n_inner_quadrature_points,
+          outer_values[0]);
+        auto &inner_values =
+          data.data_storage.get()
+            .template get_or_add_object_with_name<std::vector<ValueType>>(
+              name, data.n_inner_quadrature_points);
+        return inner_values;
+      }
+  }
+
+
+
+  template <typename ViewType>
+  template <typename VectorType>
+  const VectorType &
+  RenumberedView<ViewType>::outer_to_inner_dofs(
+    const VectorType &outer_dofs) const
+  {
+    AssertDimension(outer_dofs.size(), data.n_dofs);
+    if (data.dof_renumbering.empty())
+      {
+        return outer_dofs;
+      }
+    else
+      {
+        const auto name =
+          data.get_unique_container_name("RenumberedView::outer_to_inner_dofs",
+                                         data.n_inner_dofs,
+                                         outer_dofs[0]);
+        auto &inner_dofs = data.data_storage.get()
+                             .template get_or_add_object_with_name<VectorType>(
+                               name, data.n_inner_dofs);
+        for (unsigned int i = 0; i < data.n_dofs; ++i)
+          {
+            const auto inner_i = data.dof_renumbering[i];
+            if (inner_i != numbers::invalid_unsigned_int)
+              {
+                AssertIndexRange(inner_i, data.n_inner_dofs);
+                inner_dofs[inner_i] = outer_dofs[i];
+              }
+          }
+        return inner_dofs;
+      }
+  }
+
+
+
+  template <typename ViewType>
+  template <typename ValueType>
+  void
+  RenumberedView<ViewType>::inner_to_outer_values(
+    const std::vector<ValueType> &inner_values,
+    std::vector<ValueType>       &outer_values) const
+  {
+    AssertDimension(outer_values.size(), data.n_quadrature_points);
+    AssertDimension(inner_values.size(), data.n_inner_quadrature_points);
+    if (data.quadrature_renumbering.empty())
+      {
+        Assert(&inner_values == &outer_values, ExcInternalError());
+        return;
+      }
+    for (unsigned int i = 0; i < data.quadrature_renumbering.size(); ++i)
+      {
+        outer_values[i] = inner_values[data.quadrature_renumbering[i]];
+      }
+  }
+
+
+
+  template <typename ViewType>
+  template <typename Number>
+  void
+  RenumberedView<ViewType>::get_function_values(
+    const ReadVector<Number>                 &fe_function,
+    std::vector<solution_value_type<Number>> &values) const
+  {
+    auto &inner_values = outer_to_inner_values(values);
+    view.get_function_values(fe_function, inner_values);
+    inner_to_outer_values(inner_values, values);
```

Comment:
This too is not thread-safe. You're using a temporary vector. That vector needs to be guarded (perhaps by a different mutex than the `data.data_storage` object, or perhaps not, but it needs to be guarded).

Reply:
Same issue in the functions below.

Reply:
The whole point of using a `Threads::ThreadLocalStorage<GeneralDataStorage>` was to make this thread safe. The `outer_to_inner_values` and `inner_to_outer_values` are, in fact, both thread safe (each thread only access its own `ThreadLocalStorage` object). The vector is cached (it is created if it does not exists, and stored in the (local to thread) `GeneralDataStorage` object for future use). Why should this be not thread safe?